### PR TITLE
Revise Content-Tab »Folder« in User Mail System

### PR DESCRIPTION
**Pull request by fneumann**
See feature request:
https://docu.ilias.de/goto_docu_wiki_wpage_8177_1357.html

### CODE REVIEW

File: components/ILIAS/Mail/classes/Folder/MailFolderSearch.php [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+ * Search in mail folders
+ * - utilises database and (if active) lucene based searchers
+ * - caches results for counting of all or unread mails
+ */
+class MailFolderSearch
+{
+    private MailBoxQuery $mailbox_query;
+    private ?ilMailLuceneSearcher $lucene_searcher = null;
+    private ?ilMailSearchResult $lucene_result = null;
+    private ?array $filtered_ids = null;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Folder/MailFolderSearch.php:43]: Please try to provide the shape of the array by using PHPDoc Types: https://phpstan.org/writing-php-code/phpdoc-types

File: components/ILIAS/Mail/classes/Folder/MailFolderTableUI.php [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+                $this->getColumnDefinition(),
+                $this
+            )
+            ->withId(self::class)
+            ->withActions($this->getActions())
+            ->withRequest($this->http_request);
+    }
+
+    /**
+     * @return TableColumn[]
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Folder/MailFolderTableUI.php]:
```suggestion
     * @return array<string, TableColumn>
```

File: components/ILIAS/Mail/classes/Folder/MailFolderSearch.php [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+            $this->lucene_searcher->search($this->folder->getUserId(), $this->folder->getFolderId());
+            $this->filtered_ids = $this->lucene_result->getIds();
+        }
+        return $this->filtered_ids;
+    }
+
+    /**
+     * Inject already filtered mail ids, e.g. from a selection
+     * @param int[] $ids
+     * @return self
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Folder/MailFolderSearch.php]:
```suggestion
```

File: components/ILIAS/Mail/classes/Folder/MailFolderTableUI.php [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+                    $user->getPublicName(),
+                    (string) $this->url_builder
+                        ->withParameter($this->action_token, self::ACTION_PROFILE)
+                        ->withParameter($this->row_id_token, (string) $record->getMailId())
+                        ->buildURI()
+                ));
+            }
+            return $user->getPublicName();
+        }
+        return trim(($record->getImportName() ?? '') . ' (' . $this->lng->txt('user_deleted') . ')');
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Folder/MailFolderTableUI.php:380]:
```suggestion

        return trim(($record->getImportName() ?? '') . ' (' . $this->lng->txt('user_deleted') . ')');
```

File: components/ILIAS/Mail/classes/Folder/MailFolderTableUI.php [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+            if ($user->hasPublicProfile()) {
+                return $this->ui_renderer->render($this->ui_factory->link()->standard(
+                    $user->getPublicName(),
+                    (string) $this->url_builder
+                        ->withParameter($this->action_token, self::ACTION_PROFILE)
+                        ->withParameter($this->row_id_token, (string) $record->getMailId())
+                        ->buildURI()
+                ));
+            }
+            return $user->getPublicName();
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Folder/MailFolderTableUI.php:377]:
```suggestion

            return $user->getPublicName();
```

File: components/ILIAS/Mail/classes/Folder/MailFolderTableUI.php [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+            ? $this->ui_factory->symbol()->icon()->standard('mailr', $this->lng->txt('mailr'))
+            : $this->ui_factory->symbol()->icon()->standard('mailu', $this->lng->txt('mailu'));
+    }
+
+    private function getSender(MailRecordData $record): string
+    {
+        if ($record->getSenderId() === ANONYMOUS_USER_ID) {
+            return ilMail::_getIliasMailerName();
+        }
+        if (!empty($user = ilMailUserCache::getUserObjectById($record->getSenderId()))) {
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Folder/MailFolderTableUI.php:366]:
```suggestion

        if (!empty($user = ilMailUserCache::getUserObjectById($record->getSenderId()))) {
```

File: components/ILIAS/Mail/classes/Folder/MailFolderTableUI.php [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+            );
+        } else {
+            return sprintf(
+                '%s: %s',
+                $this->current_folder->getTitle(),
+                $this->search->getCount() === 1
+                    ? $this->lng->txt('mail_1')
+                    : sprintf($this->lng->txt('mail_s'), $this->search->getCount()),
+            );
+        }
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Folder/MailFolderTableUI.php]: You can avoid the `else` branch here, just return the string.

File: components/ILIAS/Mail/classes/Folder/MailFolderTableUI.php [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+            ),
+            self::ACTION_DELETE => $this->ui_factory->table()->action()->standard(
+                $this->lng->txt('delete'),
+                $this->url_builder->withParameter($this->action_token, self::ACTION_DELETE),
+                $this->row_id_token
+            )->withAsync(), // for confirmation modal
+        ];
+
+        foreach ($this->user_folders as $target_folder) {
+            // todo: probably check further moving restrictions (e.g. to/from drafts)
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Folder/MailFolderTableUI.php]: This seems like a forgotten TODO statement.

File: components/ILIAS/Mail/classes/Folder/MailFolderTableUI.php [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+                )
+                ->withIsSortable(true),
+        ];
+
+        if ($this->current_folder->hasOutgoingMails()) {
+            unset($columns['status'], $columns['avatar'], $columns['sender']);
+        } else {
+            unset($columns['recipients']);
+        }
+        return $columns;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Folder/MailFolderTableUI.php:161]:
```suggestion

        return $columns;
```

File: components/ILIAS/Mail/classes/Message/MailBoxOrderColumn.php [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+enum MailBoxOrderColumn: string
+{
+    case FROM = 'from';
+    case STATUS = 'm_status';
+    case SUBJECT = 'm_subject';
+    case SEND_TIME = 'send_time';
+    case RCP_TO = 'rcp_to';
+    case ATTACHMENTS = 'attachments';
+
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxOrderColumn.php]:
```suggestion
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+    private ?string $sender = null;
+    private ?string $recipients = null;
+    private ?string $subject = null;
+    private ?string $body = null;
+    private ?bool $is_unread = null;
+    private ?bool $is_system = null;
+    private ?bool $has_attachment = null;
+    private ?DateTimeImmutable $period_start = null;
+    private ?DateTimeImmutable $period_end = null;
+    private ?array $filtered_ids = null;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:52]: Please try to provide the shape of the array by using PHPDoc Types: https://phpstan.org/writing-php-code/phpdoc-types

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+    private int $limit = 999999;
+    private int $offset = 0;
+
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php]:
```suggestion
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 13]
```diff
@@ -0,0 +1,402 @@
...
+    public function withOrderColumn(?MailBoxOrderColumn $order_column): MailBoxQuery
+    {
+        $clone = clone $this;
+        if (isset($order_column)) {
+            $clone->order_column = $order_column;
+        } else {
+            $clone->order_column = self::DEFAULT_ORDER_COLUMN;
+        }
+        return $clone;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:181]:
```suggestion

        return $clone;
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 14]
```diff
@@ -0,0 +1,402 @@
...
+
+    public function withOrderDirection(?string $order_direction): MailBoxQuery
+    {
+        $clone = clone $this;
+        if (in_array($order_direction, [Order::ASC, Order::DESC])) {
+            $clone->order_direction = $order_direction;
+        } else {
+            $clone->order_direction = self::DEFAULT_ORDER_DIRECTION;
+        }
+        return $clone;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:193]:
```suggestion

        return $clone;
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 15]
```diff
@@ -0,0 +1,402 @@
...
+    }
+
+
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php]:
```suggestion
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 16]
```diff
@@ -0,0 +1,402 @@
...
+        $query = 'SELECT COUNT(m.mail_id) cnt '
+            . $this->getFrom()
+            . $this->getWhere();
+
+        $res = $this->db->query($query);
+        if ($row = $this->db->fetchAssoc($res)) {
+            return (int) $row['cnt'];
+        }
+        return 0;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:222]:
```suggestion

        return 0;
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 17]
```diff
@@ -0,0 +1,402 @@
...
+
+        if (!empty($this->period_start)) {
+            $parts[] = 'm.send_time >= ' . $this->db->quote(
+                // convert to server time zone (set by ilias initialisation)
+                $this->period_start->setTimezone(new DateTimeZone(date_default_timezone_get()))
+                                   ->format('Y-m-d H:i:s'),
+                ilDBConstants::T_TIMESTAMP
+            );
+        }
+        if (!empty($this->period_end)) {
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php]:
```suggestion

        if (!empty($this->period_end)) {
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 18]
```diff
@@ -0,0 +1,402 @@
...
+                $this->filtered_ids,
+                false,
+                ilDBConstants::T_INTEGER
+            ) . ' ';
+        }
+
+        if ($parts !== []) {
+            return ' WHERE ' . implode(' AND ', $parts);
+        }
+        return '';
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:419]:
```suggestion

        return '';
```

File: components/ILIAS/Mail/classes/class.ilMailFolderGUI.php [hunk 19]
```diff
@@ -22,29 +22,62 @@
...
     private readonly Refinery $refinery;
-    private int $currentFolderId = 0;
     private readonly ilErrorHandling $error;
-    protected readonly Factory $ui_factory;
-    protected readonly Renderer $ui_renderer;
+    private readonly Factory $ui_factory;
+    private readonly Renderer $ui_renderer;
+    private readonly ilUIService $ui_service;
+    private readonly DataFactory $data_factory;
+
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/class.ilMailFolderGUI.php]:
```suggestion
```

File: components/ILIAS/Mail/classes/Folder/MailFolderData.php [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        return !$this->isDrafts() && !$this->isSent();
+    }
+
+    public function hasOutgoingMails(): bool
+    {
+        return $this->isDrafts() || $this->isSent();
+    }
+
+
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Folder/MailFolderData.php]:
```suggestion
```

File: components/ILIAS/Mail/classes/Folder/MailFolderSearch.php [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+
+    /**
+     * Get a cached count of mails for the filter criteria
+     */
+    public function getCount(): int
+    {
+        if (!isset($this->count)) {
+            $this->count = $this->mailbox_query->withFilteredIds($this->getFilteredIds())->count();
+        }
+        return $this->count;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Folder/MailFolderSearch.php:97]:
```suggestion

        return $this->count;
```

File: components/ILIAS/Mail/classes/Folder/MailFolderSearch.php [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+
+    /**
+     * Get a cached count of unread mails for the filter criteria
+     */
+    public function getUnread(): int
+    {
+        if (!isset($this->unread)) {
+            $this->unread = $this->mailbox_query->withFilteredIds($this->getFilteredIds())->countUnread();
+        }
+        return $this->unread;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Folder/MailFolderSearch.php:109]:
```suggestion

        return $this->unread;
```

File: components/ILIAS/Mail/classes/Folder/MailFolderSearch.php [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+    private function getFilteredIds(): ?array
+    {
+        if (!isset($this->filtered_ids)
+            && isset($this->lucene_result)
+            && isset($this->lucene_searcher)
+        ) {
+            $this->lucene_searcher->search($this->folder->getUserId(), $this->folder->getFolderId());
+            $this->filtered_ids = $this->lucene_result->getIds();
+        }
+        return $this->filtered_ids;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Folder/MailFolderSearch.php:167]:
```suggestion

        return $this->filtered_ids;
```

File: components/ILIAS/Mail/classes/Folder/MailFolderTableUI.php [hunk 24]
```diff
@@ -0,0 +1,408 @@
...
+
+    private function getAvatar(MailRecordData $record): string
+    {
+        if (!array_key_exists($record->getSenderId(), $this->avatars)) {
+            $user = ilMailUserCache::getUserObjectById($record->getSenderId());
+            $this->avatars[$record->getSenderId()] = isset($user)
+                ? $this->ui_renderer->render($user->getAvatar())
+                : '';
+        }
+        return $this->avatars[$record->getSenderId()];
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Folder/MailFolderTableUI.php:350]:
```suggestion

        return $this->avatars[$record->getSenderId()];
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 25]
```diff
@@ -0,0 +1,402 @@
...
+        $query = 'SELECT m.mail_id '
+            . $this->getFrom()
+            . $this->getWhere();
+
+        $ids = [];
+        $res = $this->db->query($query);
+        while ($row = $this->db->fetchAssoc($res)) {
+            $ids[] = (int) $row['mail_id'];
+        }
+        return $ids;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:245]:
```suggestion

        return $ids;
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 26]
```diff
@@ -0,0 +1,402 @@
...
+                isset($row['rcp_bcc']) ? (string) $row['rcp_bcc'] : null,
+                isset($row['attachments']) ? (array) unserialize(
+                    stripslashes($row['attachments']),
+                    ['allowed_classes' => false]
+                ) : [],
+                isset($row['tpl_ctx_id']) ? (string) $row['tpl_ctx_id'] : null,
+                isset($row['tpl_ctx_params']) ? (string) $row['tpl_ctx_params'] : null
+            );
+        }
+        return $set;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:308]:
```suggestion

        return $set;
```

File: components/ILIAS/Mail/classes/class.ilMailFolderGUI.php [hunk 27]
```diff
@@ -186,304 +354,256 @@ protected function showUser(): void
...
-        $mailIds = $this->getMailIdsFromRequest();
-        if ($mailIds !== []) {
-            $this->umail->markUnread($mailIds);
-            $this->tpl->setOnScreenMessage('success', $this->lng->txt('saved_successfully'));
-        } else {
-            $this->tpl->setOnScreenMessage('info', $this->lng->txt('mail_select_one'));
-        }
-
-        $this->showFolder();
+        return [];
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/class.ilMailFolderGUI.php:604]:
```suggestion

        return [];
```

File: components/ILIAS/Mail/classes/class.ilMailFolderGUI.php [hunk 28]
```diff
@@ -515,82 +635,64 @@ protected function moveSingleMail(): void
...
-            $this->ctrl->redirect($this, 'showFolder');
-        }
+        $modal = $this->ui_factory->modal()->interruptive(
+            $this->lng->txt('delete'),
+            $this->lng->txt('mail_sure_delete'),
+            $this->ctrl->getFormAction($this, self::CMD_DELETE_MAILS)
+        )->withAffectedItems($items);
 
-        $this->showFolder();
+        echo $this->ui_renderer->renderAsync($modal);
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/class.ilMailFolderGUI.php]: Please use the HTTP service for "echo/exit" use cases. Don't hesitate to contact me if you have any questions.

File: components/ILIAS/Mail/classes/class.ilMailFolderGUI.php [hunk 29]
```diff
@@ -61,110 +94,245 @@ public function __construct()
...
+            // no redirect possible from async call
+            if ($action === MailFolderTableUI::ACTION_DELETE) {
+                $modal = $this->ui_factory->modal()->lightbox(
+                    $this->ui_factory->modal()->lightboxTextPage(
+                        $this->lng->txt('mail_select_one'),
+                        $this->lng->txt('delete'),
+                    )
+                );
+                echo $this->ui_renderer->render($modal);
+                exit;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/class.ilMailFolderGUI.php]: Please use the HTTP service for "echo/exit" use cases.

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 30]
```diff
@@ -0,0 +1,402 @@
...
+    ) {
+        global $DIC;
+        $this->db = $DIC->database();
+    }
+
+    public function withFolderId(?int $folder_id): MailBoxQuery
+    {
+        $clone = clone $this;
+        $clone->folder_id = $folder_id;
+        return $clone;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:70]:
```suggestion

        return $clone;
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 31]
```diff
@@ -0,0 +1,402 @@
...
+    }
+
+    public function withSender(?string $sender): MailBoxQuery
+    {
+        $clone = clone $this;
+        $clone->sender = $sender;
+        return $clone;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:78]:
```suggestion

        return $clone;
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 32]
```diff
@@ -0,0 +1,402 @@
...
+    }
+
+    public function withRecipients(?string $recipients): MailBoxQuery
+    {
+        $clone = clone $this;
+        $clone->recipients = $recipients;
+        return $clone;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:86]:
```suggestion

        return $clone;
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 33]
```diff
@@ -0,0 +1,402 @@
...
+    }
+
+    public function withSubject(?string $subject): MailBoxQuery
+    {
+        $clone = clone $this;
+        $clone->subject = $subject;
+        return $clone;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:94]:
```suggestion

        return $clone;
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 34]
```diff
@@ -0,0 +1,402 @@
...
+    }
+
+    public function withBody(?string $body): MailBoxQuery
+    {
+        $clone = clone $this;
+        $clone->body = $body;
+        return $clone;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:102]:
```suggestion

        return $clone;
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 35]
```diff
@@ -0,0 +1,402 @@
...
+    }
+
+    public function withIsUnread(?bool $is_unread): MailBoxQuery
+    {
+        $clone = clone $this;
+        $clone->is_unread = $is_unread;
+        return $clone;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:110]:
```suggestion

        return $clone;
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 36]
```diff
@@ -0,0 +1,402 @@
...
+    }
+
+    public function withIsSystem(?bool $is_system): MailBoxQuery
+    {
+        $clone = clone $this;
+        $clone->is_system = $is_system;
+        return $clone;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:118]:
```suggestion

        return $clone;
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 37]
```diff
@@ -0,0 +1,402 @@
...
+    }
+
+    public function withHasAttachment(?bool $has_attachment): MailBoxQuery
+    {
+        $clone = clone $this;
+        $clone->has_attachment = $has_attachment;
+        return $clone;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:126]:
```suggestion

        return $clone;
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 38]
```diff
@@ -0,0 +1,402 @@
...
+    }
+
+    public function withPeriodStart(?DateTimeImmutable $period_start): MailBoxQuery
+    {
+        $clone = clone $this;
+        $clone->period_start = $period_start;
+        return $clone;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:134]:
```suggestion

        return $clone;
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 39]
```diff
@@ -0,0 +1,402 @@
...
+    }
+
+    public function withPeriodEnd(?DateTimeImmutable $period_end): MailBoxQuery
+    {
+        $clone = clone $this;
+        $clone->period_end = $period_end;
+        return $clone;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:142]:
```suggestion

        return $clone;
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 40]
```diff
@@ -0,0 +1,402 @@
...
+    }
+
+    /**
+     * @param int[]|null $filtered_ids
+     */
+    public function withFilteredIds(?array $filtered_ids): MailBoxQuery
+    {
+        $clone = clone $this;
+        $clone->filtered_ids = $filtered_ids;
+        return $clone;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:153]:
```suggestion

        return $clone;
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 41]
```diff
@@ -0,0 +1,402 @@
...
+    }
+
+    public function withLimit(int $limit): MailBoxQuery
+    {
+        $clone = clone $this;
+        $clone->limit = $limit;
+        return $clone;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:161]:
```suggestion

        return $clone;
```

File: components/ILIAS/Mail/classes/Message/MailBoxQuery.php [hunk 42]
```diff
@@ -0,0 +1,402 @@
...
+    }
+
+    public function withOffset(int $offset): MailBoxQuery
+    {
+        $clone = clone $this;
+        $clone->offset = $offset;
+        return $clone;
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Message/MailBoxQuery.php:169]:
```suggestion

        return $clone;
```

File: components/ILIAS/Mail/classes/class.ilMailFolderGUI.php [hunk 43]
```diff
@@ -186,304 +354,256 @@ protected function showUser(): void
...
 
-        $table_html = $mailtable->getHTML();
-
-        if (!$oneConfirmationDialogueRendered && !$this->confirmTrashDeletion) {
-            $this->toolbar->setFormAction($this->ctrl->getFormAction($this, 'showFolder'));
+        [   $url_builder,
+            $action_token,
+            $row_id_token,
+            $target_token,
+        ] = (new URLBuilder($this->data_factory->uri(
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/class.ilMailFolderGUI.php]:
```suggestion
        [
            $url_builder,
            $action_token,
            $row_id_token,
            $target_token,
        ] = (new URLBuilder($this->data_factory->uri(
```

File: components/ILIAS/Mail/classes/Folder/MailFolderSearch.php [hunk 44]
```diff
@@ -0,0 +1,178 @@
...
+    /**
+     * Get the cached mail ids from a lucene search for selected filter criteria
+     * These will be used as additional filter for the mailbox query
+     */
+    private function getFilteredIds(): ?array
+    {
+        if (!isset($this->filtered_ids)
+            && isset($this->lucene_result)
+            && isset($this->lucene_searcher)
+        ) {
```
Comment (mjansenDatabay) [inline components/ILIAS/Mail/classes/Folder/MailFolderSearch.php]:
```suggestion
        if (!isset($this->filtered_ids) &&
            isset($this->lucene_result) &&
            isset($this->lucene_searcher)) {
```

**Review by mjansenDatabay — CHANGES_REQUESTED** [review]
Hi @fneumann ,

thx for your excellent work. I only found a few minor issues and added some code style remarks (because of personal preference).

Please apply the following changes:
- [x] Replace usages of `isset` on member variables of classes with a `$this->variable !== null` check, or `!isset` with `$this->variable === null`.

And one question:
- What was the personal indicator for you to introduce `Enum` for some use cases and constants for other use cases? I am just curious, I don't see any necessity to change this..

Best regards
Michael

**Comment by fneumann** [comment]
Resolved all threads. 
Fixed an issue with unrecognized '0' as a filter condition. 

Regarding Enums and class constant:
I created Enums to use them as typed function parameters or return values. The actions of MailFolderTableUI are never passed this way but only used internally as array keys and externally as query parameters. Having them as class constants seemed enough.

**Pull request merged**